client/asset/btc/btc.go

```diff
@@ -1747,336 +1733,206 @@ func (btc *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, time.Time,
+	for iHeight <= tipHeight {
+		validReqs := make(map[outPoint]*findRedemptionReq, len(undiscovered))
+		for outPt, req := range undiscovered {
+			if iHeight >= req.blockHeight {
+				validReqs[outPt] = req
 			}
 		}
-		btc.findRedemptionMtx.RUnlock()
-		btc.findRedemptionsInMempool(pendingContracts)
-	}
-}
-
-// findRedemptionsInTx checks if any input of the passed tx spends any of the
-// specified contract outpoints. If spending info is found for any contract, the
-// contract's secret or any error encountered while trying to parse the secret
-// is returned to the redemption finder via the registered result chan; and the
-// contract is purged from the findRedemptionQueue.
-// Returns the number of redemptions found and canceled.
-func (btc *ExchangeWallet) findRedemptionsInTx(scanPoint string, tx *btcjson.TxRawResult, contractOutpoints []outPoint) (found, cancelled int) {
-	btc.findRedemptionMtx.Lock()
-	defer btc.findRedemptionMtx.Unlock()
 
-	extractSecret := func(vin int, contractHash []byte) (*chainhash.Hash, []byte, error) {
-		redeemTxHash, err := chainhash.NewHashFromStr(tx.Txid)
-		if err != nil {
-			return nil, nil, err
-		}
-		if tx.Vin[vin].ScriptSig == nil {
-			return nil, nil, fmt.Errorf("no sigScript")
-		}
-		sigScript, err := hex.DecodeString(tx.Vin[vin].ScriptSig.Hex)
-		if err != nil {
-			return nil, nil, err
-		}
-		var witness [][]byte
-		for _, hexB := range tx.Vin[vin].Witness {
-			var b []byte
-			b, err = hex.DecodeString(hexB)
-			if err != nil {
-				break
+		discovered := btc.node.searchBlockForRedemptions(validReqs, *iHash)
+		for outPt, res := range discovered {
+			req, found := undiscovered[outPt]
+			if !found {
+				btc.log.Errorf("Request not found in undiscovered map. This shouldn't be possible.")
+				continue
 			}
-			witness = append(witness, b)
-		}
-		txIn := wire.NewTxIn(new(wire.OutPoint), sigScript, witness)
-		secret, err := dexbtc.FindKeyPush(txIn, contractHash, btc.segwit, btc.chainParams)
-		if err != nil {
-			return nil, nil, err
+			req.resultChan <- res
+			delete(undiscovered, outPt)
 		}
-		return redeemTxHash, secret, nil
-	}
 
-	for _, contractOutpoint := range contractOutpoints {
-		req, exists := btc.findRedemptionQueue[contractOutpoint]
-		if !exists {
-			continue // no find request for this outpoint (impossible now?)
-		}
-		if req.canceled() {
-			cancelled++
-			delete(btc.findRedemptionQueue, contractOutpoint)
-			continue // this find request has been cancelled
+		if len(undiscovered) == 0 {
+			return
 		}
 
-		for i := range tx.Vin {
-			input := &tx.Vin[i]
-			if input.Vout != contractOutpoint.vout || input.Txid != contractOutpoint.txHash.String() {
-				continue // input doesn't redeem this contract, check next input
+		if len(discovered) > 0 {
+			// If we found some, we may be able to skip some blocks
+			lowHash, lowHeight := lowestBlock()
+			if lowHeight >= iHeight+1 {
+				iHash, iHeight = lowHash, lowHeight
+				continue
```

Comment:
Related to a previous comment. I can't think up a scenario where any contract's block is higher than the start block height. This code block won't be hit if there's just one request to process and the redemption has been found for that request. If there are multiple requests, then the start block is most likely the best block at the time the search was started. The contracts being searched for then could not have a block height higher than the best block height.